# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


In [22]:
import pandas as pd
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
car_customers = pd.read_csv(url)

External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [23]:
#Identify the dimensions of the dataset by determining the number of rows and columns it contains.

car_customers.shape

(4008, 11)

In [29]:
#Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. 
#You should also provide suggestions for fixing any incorrect data types

data_types = car_customers.dtypes
print(data_types)

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


In [30]:
#Identify the number of unique values for each column and determine which columns appear to be categorical. 


unique_counts = car_customers.nunique()
print(unique_counts)


Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64


Customer: is a unique identifier with 1071 unique values
ST (State): has 8 unique values - meaning presence in 8 different states - CATEGORICAL
GENDER: has 5 unique values, suggesting 5 different gender categories - CATEGORICAL
Education: has 6 unique values, indicating 6 different education levels - CATEGORICAL
Customer Lifetime Value: has 1027 unique values
Income: has 774 unique values, indicating a range of income levels
Monthly Premium Auto: has 132 unique values
Number of Open Complaints: has 6 unique values, representing the number of open complaints
Policy Type: has 3 unique values, suggesting 3 different policy types - CATEGORICAL
Vehicle Class: has 6 unique values, indicating various vehicle classes - CATEGORICAL
Total Claim Amount: has 761 unique values

In [31]:
#You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
unique_states = car_customers['ST'].unique()
print("Unique values in 'ST' (State) column:")
print(unique_states)


Unique values in 'ST' (State) column:
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]


In [37]:
state_mapping = {
    'WA': 'Washington',
    'AZ': 'Arizona',
    'CA': 'California',
    'NV': 'Nevada',
    'OR': 'Oregon',
    'Cali': 'California',  # Map 'Cali' to 'California'
}

car_customers['ST'] = car_customers['ST'].replace(state_mapping).fillna('Unknown')

unique_states = car_customers['ST'].unique()
print("Unique values in 'ST' (State) column after standardization:")
print(unique_states)



Unique values in 'ST' (State) column after standardization:
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Unknown']


In [39]:
unique_gender = car_customers['GENDER'].unique()
print("Unique values in 'GENDER' column:")
print(unique_gender)



Unique values in 'GENDER' column:
[nan 'F' 'M' 'Femal' 'Male' 'female']


In [40]:
gender_mapping = {
    'Femal': 'Female',
    'female': 'Female',
    'Male': 'Male',
}

car_customers['GENDER'] = car_customers['GENDER'].replace(gender_mapping).fillna('Unknown')

unique_gender = car_customers['GENDER'].unique()
print("Unique values in 'GENDER' column after standardization:")
print(unique_gender)


Unique values in 'GENDER' column after standardization:
['Unknown' 'F' 'M' 'Female' 'Male']


In [41]:
unique_education = car_customers['Education'].unique()
print("Unique values in 'Education' column:")
print(unique_education)


Unique values in 'Education' column:
['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]


In [42]:
education_mapping = {
    'Bachelors': 'Bachelor',
    'Bachelors': 'Bachelor',
    'Master': 'Master',
    'Doctor': 'Doctor',
    'High School or Below': 'High School',  
}
car_customers['Education'] = car_customers['Education'].replace(education_mapping).fillna('Unknown')

unique_education = car_customers['Education'].unique()
print("Unique values in 'Education' column after standardization:")
print(unique_education)


Unique values in 'Education' column after standardization:
['Master' 'Bachelor' 'High School' 'College' 'Doctor' 'Unknown']


In [43]:
unique_policy_type = car_customers['Policy Type'].unique()
print("Unique values in 'Policy Type' column:")
print(unique_policy_type)


Unique values in 'Policy Type' column:
['Personal Auto' 'Corporate Auto' 'Special Auto' nan]


In [44]:
unique_vehicle_class = car_customers['Vehicle Class'].unique()
print("Unique values in 'Vehicle Class' column:")
print(unique_vehicle_class)


Unique values in 'Vehicle Class' column:
['Four-Door Car' 'Two-Door Car' 'SUV' 'Luxury SUV' 'Sports Car'
 'Luxury Car' nan]


Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. 

In [45]:
numerical_summary = car_customers.describe()
print(numerical_summary)


       Customer Lifetime Value        Income  Monthly Premium Auto  \
count             1.068000e+03   1071.000000           1071.000000   
mean              7.936903e+05  39295.701214            193.234360   
std               6.434784e+05  30469.427060           1601.190369   
min               2.004351e+05      0.000000             61.000000   
25%               4.034080e+05  14072.000000             68.000000   
50%               5.881742e+05  36234.000000             83.000000   
75%               8.962872e+05  64631.000000            109.500000   
max               5.816655e+06  99960.000000          35354.000000   

       Total Claim Amount  
count         1071.000000  
mean           404.986909  
std            293.027260  
min              0.382107  
25%            202.157702  
50%            354.729129  
75%            532.800000  
max           2893.239678  


You should also provide your conclusions based on these summary statistics.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [38]:
# Create a Series with customer locations and their frequencies
customer_locations_series = car_customers['ST'].value_counts()

# Sort the Series in ascending order
customer_locations_sorted = customer_locations_series.sort_values()

# Retrieve the top 5 less common locations
top_5_less_common = customer_locations_sorted.head(5)

print("Top 5 Less Common Customer Locations (Ascending Order):")
print(top_5_less_common)


Top 5 Less Common Customer Locations (Ascending Order):
ST
Nevada         98
Washington    111
Arizona       211
Oregon        320
California    331
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [46]:
# Your code here# Use value_counts to count the number of policies sold for each policy type
policy_counts = car_customers['Policy Type'].value_counts()

# Find the policy type with the highest number of policies sold
most_sold_policy = policy_counts.idxmax()

print("Total number of policies sold for each policy type:")
print(policy_counts)

print("Policy type with the highest number of policies sold:", most_sold_policy)


Total number of policies sold for each policy type:
Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64
Policy type with the highest number of policies sold: Personal Auto


### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [47]:
# Create a DataFrame for Personal Auto policies
personal_auto_df = car_customers.loc[car_customers['Policy Type'] == 'Personal Auto']

# Create a DataFrame for Corporate Auto policies
corporate_auto_df = car_customers.loc[car_customers['Policy Type'] == 'Corporate Auto']

# Calculate the average income for each policy type
average_income_personal_auto = personal_auto_df['Income'].mean()
average_income_corporate_auto = corporate_auto_df['Income'].mean()

# Print the results
print("Average income for Personal Auto policies:", average_income_personal_auto)
print("Average income for Corporate Auto policies:", average_income_corporate_auto)


Average income for Personal Auto policies: 38180.69871794872
Average income for Corporate Auto policies: 41390.31196581197


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here